pkg/ring/replication_set.go

```diff
@@ -13,23 +14,37 @@ type ReplicationSet struct {
 
 // Do function f in parallel for all replicas in the set, erroring is we exceed
 // MaxErrors and returning early otherwise.
-func (r ReplicationSet) Do(ctx context.Context, f func(*IngesterDesc) (interface{}, error)) ([]interface{}, error) {
+func (r ReplicationSet) Do(ctx context.Context, delay time.Duration, f func(*IngesterDesc) (interface{}, error)) ([]interface{}, error) {
 	errs := make(chan error, len(r.Ingesters))
 	resultsChan := make(chan interface{}, len(r.Ingesters))
+	minSuccess := len(r.Ingesters) - r.MaxErrors
+	done := make(chan struct{})
+	forceStart := make(chan struct{}, r.MaxErrors)
+	defer func() {
+		close(done)
+	}()
 
-	for _, ing := range r.Ingesters {
-		go func(ing *IngesterDesc) {
+	for i, ing := range r.Ingesters {
+		go func(i int, ing *IngesterDesc) {
+			// wait to send extra requests
+			if i >= minSuccess && delay > 0 {
+				select {
+				case <-done:
+					return
+				case <-forceStart:
+				case <-time.After(delay):
```

Comment:
Wondering if there should be a request-scoped `Timer` here, shared between all the parallel downstream requests?  Probably not worth it TBH.

Reply:
It would be a tiny bit more efficient, but at the cost of readability of the code. May need to add yet another channel anyway to trigger all of the calls since we can't actually close a Timer's channel. I vote not worth it.